### PR TITLE
Add UpgradeOnMovement trait

### PR DESCRIPTION
**Pull request by reaperrr**
As the name states, the trait provides upgrades while an actor is moving, this can be used to increase inaccuracy while moving, give the unit a more vulnerable armor type while moving and so on.

This nicely complements our set of upgrades-providing traits.

The trait is upgradable itself, which could be used to switch the provided upgrades manually in conjunction with `DeployToUpgrade`.

### CODE REVIEW

**Comment by ABrandau** [comment]
Seems to work fine, I emped the unit to see if it had any issues and nope, same as when the unit attack moves.

**Comment by WolfGaming** [comment]
Do we want to have a property that flips the functionality? Aka grants an upgrade when not moving?

**Comment by pchote** [comment]
The upgrades system already lets you do that without making a new type.

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		public override object Create(ActorInitializer init) { return new UpgradeOnMovement(init.Self, this); }
+	}
+
+	public class UpgradeOnMovement : UpgradableTrait<UpgradeOnMovementInfo>, ITick
+	{
+		readonly UpgradeManager manager;
+		readonly IMove movement;
+
+		bool granted;
+		bool isMoving;
```
Comment (WolfGaming) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs]: Why have this here and not defined in the tick like oldCachedPosition?

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+		void ITick.Tick(Actor self)
+		{
+			if (IsTraitDisabled)
+				return;
+
+			var oldCachedPosition = cachedPosition;
+			cachedPosition = self.CenterPosition;
+
+			isMoving = !self.IsDead && (movement.IsMoving || (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0);
```
Comment (WolfGaming) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs]: Do you do (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0 for actors inside transports or am I missing something?

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs]: During my tests there were some 'hiccups' when only using IMove.IsMoving (during turns, usually). If we managed to make that more robust instead this could go away, but until then this is needed to make sure the upgrades are applied all the time while moving.

**Comment by reaperrr** [comment]
Updated:
- upgrades are now also revoked if `UpgradeOnMovement` is disabled
- `isMovement` is now a `var` inside Tick
- moved the testcase to https://github.com/reaperrr/OpenRA/tree/UpgradeOnMovement-test

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+			if (IsTraitDisabled)
+			{
+				Revoke(self);
+				return;
+			}
+
+			var oldCachedPosition = cachedPosition;
+			cachedPosition = self.CenterPosition;
+
+			var isMoving = !self.IsDead && (movement.IsMoving || (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs]: This "breaks" with aircraft. They always have `movement.IsMoving = self.World.Map.DistanceAboveTerrain(CenterPosition).Length > 0`.

Question now: Do we want to have this disabled for hovering helicopters (I'd say yes).
This could also be left to another PR, as it seems to be a general "bug" with aircraft.

**Comment by abcdefg30** [comment]
Looks good to me otherwise. :+1: / :white_check_mark:

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs (hunk 3 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs]: So, should I change the `Aircraft.IsMoving` check to only return `true` if the aircraft is moving (horizontally or vertically)?

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs]: For at least the orca `WithMoveAnimation` case we'll only want to consider horizontal movements.  I believe that we've already ported all the other fiddly uses of `Aircraft.IsMoving` to upgrades so I *think* (but please check) we're otherwise safe to make that change.

**Comment by reaperrr** [comment]
Updated and rebased.

I didn't update the testcase, instead the TD Orca now uses this instead of `WithMoveAnimation`.
Note: I have an upgrade rule ready to get rid of `WithMoveAnimation`, but we need to make `WithAttackAnimation` able to handle multiple sprite body traits before we can use this for actors that use both (like the TS Wolverine).

@abcdefg30 could you revalidate your +1?

**Comment by pchote** [comment]
We need to complete rewrite/rethink the way `WithInfantryBody` etc work with respect to animations, the current approach just doesn't work (and changing WithAttackAnimation etc won't help).  The body trait needs to listen to these things itself so that it can handle the transitions between them (e.g. tiltrotor animations, or liedown/standup).

**Comment by reaperrr** [comment]
Generally speaking I agree, but that sounds like another larger-scale refactor and those usually take time.

**Comment by abcdefg30** [comment]
Looks still good to me. 👍 / ✅

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+namespace OpenRA.Mods.Common.Traits
+{
+	public class UpgradeOnMovementInfo : UpgradableTraitInfo, Requires<UpgradeManagerInfo>, Requires<IMoveInfo>
+	{
+		[UpgradeGrantedReference, FieldLoader.Require]
+		[Desc("The upgrades to grant.")]
+		public readonly string[] Upgrades = { };
+
+		[Desc("Apply upgrades on straight vertical movement as well.")]
+		public readonly bool ConsiderStraightVerticalMovement = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs]: I find this naming confusing.  Lets simplify this to `ConsiderVerticalMovement`?

File: mods/cnc/rules/aircraft.yaml [hunk 5]
```diff
@@ -158,8 +158,15 @@ ORCA:
 	Explodes:
 		Weapon: HeliExplode
 		EmptyWeapon: HeliExplode
-	WithMoveAnimation:
-		MoveSequence: move
+	UpgradeOnMovement:
+		Upgrades: movingforward
+	WithFacingSpriteBody:
```
Comment (pchote) [inline mods/cnc/rules/aircraft.yaml]: Given our more recent discovery / discussion about animations and sprite bodies I think it might be a better idea to keep the Orca as it is, keeping this trait for downstream modders.

**Review by pchote — CHANGES_REQUESTED** [review]
Code changes look fine, and works as advertised so :+1: after we resolve the following.

**Comment by reaperrr** [comment]
Updated as requested.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5e4566ddb90509af8d946ebcd1d028a8dd0a43bd
